api/src/main/java/io/opencensus/trace/Tracestate.java

```diff
@@ -0,0 +1,255 @@
+      Entry entry = Entry.create(key, value);
+      if (toAdd == null) {
+        // Copy entries from the parent.
+        toAdd = new ArrayList<Entry>(parent.getEntries());
+      }
+      for (int i = 0; i < toAdd.size(); i++) {
+        if (toAdd.get(i).getKey().equals(entry.getKey())) {
+          toAdd.remove(i);
+          // Exit now because the toAdd list cannot contain duplicates.
+          break;
+        }
+      }
+      // Inserts the element at the front of this list.
+      toAdd.add(0, entry);
+      return this;
+    }
+
+    /**
+     * Removes the {@code Entry} that has the given {@code key} if it is present.
+     *
+     * @param key the key for the {@code Entry} to be removed.
+     * @return this.
+     * @since 0.16
+     */
+    public Builder remove(String key) {
+      Utils.checkNotNull(key, "key");
+      if (toAdd == null) {
+        // Copy entries from the parent.
+        toAdd = new ArrayList<Entry>(parent.getEntries());
+      }
+      for (int i = 0; i < toAdd.size(); i++) {
+        if (toAdd.get(i).getKey().equals(key)) {
+          toAdd.remove(i);
+          // Exit now because the toAdd list cannot contain duplicates.
+          break;
+        }
+      }
+      return this;
+    }
+
+    /**
+     * Builds a TraceState by adding the entries to the parent in front of the key-value pairs list
+     * and removing duplicate entries.
+     *
+     * @return a TraceState with the new entries.
+     * @since 0.16
+     */
+    public Tracestate build() {
+      if (toAdd == null) {
+        return parent;
+      }
+      return Tracestate.create(toAdd);
+    }
+  }
+
+  /**
+   * Immutable key-value pair for {@code Tracestate}.
+   *
+   * @since 0.16
+   */
+  @Immutable
+  @AutoValue
+  public abstract static class Entry {
+    /**
+     * Creates a new {@code Entry} for the {@code Tracestate}.
+     *
+     * @param key the Entry's key.
+     * @param value the Entry's value.
+     * @since 0.16
+     */
+    public static Entry create(String key, String value) {
+      Utils.checkNotNull(key, "key");
+      Utils.checkNotNull(value, "value");
+      Utils.checkArgument(validateKey(key), "Invalid key " + key);
+      Utils.checkArgument(validateValue(value), "Invalid value " + value);
+      return new AutoValue_Tracestate_Entry(key, value);
+    }
+
+    /**
+     * Returns the key {@code String}.
+     *
+     * @return the key {@code String}.
+     * @since 0.16
+     */
+    public abstract String getKey();
+
+    /**
+     * Returns the value {@code String}.
+     *
+     * @return the value {@code String}.
+     * @since 0.16
+     */
+    public abstract String getValue();
+
+    Entry() {}
+  }
+
+  // Key is opaque string up to 256 characters printable. It MUST begin with a lowercase letter, and
+  // can only contain lowercase letters a-z, digits 0-9, underscores _, dashes -, asterisks *, and
+  // forward slashes /.
+  private static boolean validateKey(String key) {
```

Comment:
should this be public to allow other protocol's implementations to use this method?

Reply:
can you link us on protocol differences? seems important if they are more constrained not less as most constrained should win

Reply:
I would not make it public now. We can always change that when we see a need.